### PR TITLE
feat(oauth-login): Basic oauth implementation

### PR DESCRIPTION
**Pull request by cedricgrothues**
This PR adds basic `Login with GitHub` functionality.

### CODE REVIEW

File: server/src/main.rs [hunk 1]
```diff
@@ -1,4 +1,81 @@
...
+    let client = reqwest::Client::new();
+
+    if let Ok(res) = client
+        .post("https://github.com/login/oauth/access_token")
+        .header("accept", "application/json")
+        .form(&params)
+        .send()
+        .await
+    {
+        return Ok(format!("{:#?}", res.json::<TokenResponse>().await.unwrap()));
```
Comment (teymour-aldridge) [inline server/src/main.rs:77]: I think we should add the access token to the database.

Comment (cedricgrothues) [reply server/src/main.rs:77]: Yeah, that was just for debugging. I should also get rid of that `unwrap` 😨

Comment (teymour-aldridge) [reply server/src/main.rs:77]: Sounds good. I reckon if we rebase of master then it should be pretty painless.

File: server/src/main.rs [hunk 2]
```diff
@@ -1,6 +1,196 @@
...
+                            .values(data::NewPod {
+                                pod_id: pods[0].name.clone(),
+                                pod_name: pods[0].description.clone(),
+                            })
+                            .returning(data::pod::id)
+                            .get_result::<i32>(&d.get().unwrap())
+                        {
+                            Ok(r) => Ok(r),
+                            Err(_) => {
+                                let r = match data::pod::dsl::pod
```
Comment (teymour-aldridge) [inline server/src/main.rs:137]: Neat! (although the variable names are a bit hard to understand)

File: server/src/main.rs [hunk 3]
```diff
@@ -1,6 +1,196 @@
...
+
+                    if pod_id.is_err() {
+                        let response = Response::builder()
+                            .status(424)
+                            .body("error: failed dependency (pod_id)");
+
+                        return Ok(response);
+                    }
+
+                    let _ = diesel::insert_into(data::author::dsl::author)
```
Comment (teymour-aldridge) [inline server/src/main.rs:157]: Anything bound to `_` is immediately dropped so I don't think this is needed.
```suggestion
                    diesel::insert_into(data::author::dsl::author)
```

File: server/src/main.rs [hunk 4]
```diff
@@ -1,6 +1,196 @@
...
+                        .values(data::NewAuthor {
+                            access_token: token.access_token.clone(),
+                            github_username: query.data.viewer.login,
+                            email: query.data.viewer.email,
+                            pod_id: pod_id.unwrap(),
+                            role: AuthorRole::Fellow,
+                            status: AuthorStatus::Good,
+                        })
+                        .returning((data::author::email, data::author::github_username))
```
Comment (teymour-aldridge) [inline server/src/main.rs:166]: The suggestion above and the suggestion below should be taken as one suggestion – I'm not sure if you intend to do something with the data here, but if you don't I don't think we need it.
```suggestion
```

File: server/src/main.rs [hunk 5]
```diff
@@ -1,6 +1,196 @@
...
+                        .get_result::<(String, String)>(&d.get().unwrap());
```
Comment (teymour-aldridge) [inline server/src/main.rs:167]:
```suggestion
                        .execute(&d.get().unwrap());
```

File: server/src/main.rs (hunk 3 above)
Comment (cedricgrothues) [reply server/src/main.rs:157]: The `_` prevents a `Unhandled Result` warning

Comment (teymour-aldridge) [reply server/src/main.rs:157]: That's interesting – not sure why! Maybe something to investigate after the hackathon.

File: server/src/main.rs [hunk 6]
```diff
@@ -1,6 +1,195 @@
...
+    let public = warp::get().and(warp::fs::dir("./server/public"));
+
+    let redirect_url = Uri::from_str(&format!(
+        "https://github.com/login/oauth/authorize?client_id={}&scope=read:org,read:user&state=random",
+        CLIENT_ID.to_string()
+    ))
+    .unwrap();
+
+    let database = crate::db::Pool::new(ConnectionManager::new(
+        "postgres://postgres@localhost/halfway-hackathon",
```
Comment (teymour-aldridge) [inline server/src/main.rs:33]:
```suggestion
        var("DATABASE_URL").expect("The `DATABASE_URL` environment variable is not set")
```
I think we should ideally initialize this from an environment variable for configurability, after which this should be good to merge.

File: server/src/main.rs [hunk 7]
```diff
@@ -132,14 +132,14 @@ async fn callback(d: db::Pool, q: CallbackQuery) -> Result<impl warp::Reply, Inf
...
                                     .filter(data::pod::pod_id.eq(pods[0].name.clone()))
                                     .load::<(i32, String, String)>(&d.get().unwrap())
                                 {
-                                    Ok(r) => Ok(r[0].0),
-                                    Err(e) => Err(e),
+                                    Ok(result) => Ok(result[0].0),
+                                    Err(error) => Err(error),
                                 };
 
                                 r
```
Comment (teymour-aldridge) [inline server/src/main.rs:145]:
```suggestion
                                result
```

File: server/src/main.rs (hunk 6 above)
Comment (cedricgrothues) [reply server/src/main.rs:33]: You're right, I forgot to remove that 🤦‍♂️

**Pull request merged**